ipwb/replay.py

```diff
@@ -253,8 +249,22 @@ def resolveMemento(urir, datetime):
     return (newDatetime, linkHeader, uri)
 
 
+def compile_target_uri(url: str, query_string: bytes) -> str:
```

Comment:
Is this `compile_target_uri` function any better, more accurate, or more efficient than previously defined `getCompleteURI`?

Reply:
1. Missing query string is not a bug of Flask, contrary to the comment at `def getCompleteURI(uri)` - that's how Flask works. Thus, I moved the call of this function from the internals of `ipwb`'s business logic into Flask handlers. Business logic should not care about such things.

2. `getCompleteURI` is not a pure function since it relies upon global `flask.request` singleton. That makes it hard to test. `compile_target_uri` does not have that problem, and unit tests for it are in this PR.

3. I confess I could not resist the urge to change the piece of code I was modifying a bit towards Python's conventions. One of those conventions is:

> Function names should be lowercase, with words separated by underscores as necessary to improve readability.
https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

That's why I renamed the function and, as a bonus, added type annotations.

I am also aware that all the other functions in the project are named in camelCase, more in Java tradition, and my change breaks consistency, which is bad. I can make a PR to rename all the other functions from camelCase to snake_case if you'd be open to that.

4. New function is a bit more verbose, but I also believe it is easier to understand.

Reply:
@anatoly-scherbakov Thanks for your contributions, the code needs more eyes for sanity checks by those more familiar with conventions (naming) and imported modules' base functionality (e.g., Flask).

I would be open to overhauling the naming if it puts us more inline with Python standards, which might in turn facilitate further contributions by those already familiar with the style.

Reply:
Thanks @anatoly-scherbakov for putting forward your rationale. These camelCase function names are footprints of early code, which Mat was gradually changing over time. I think the most strong argument from you is the purity of the function which allows reliable unit testing.

Your `str.format()` calls are more verbose than they need to be. They are not bad, but I just find it way too verbose to to define non-reusable variables and assigning values to them and immediately using them. A string of joust a couple of pieces can perhaps be composed using positional placeholders without the fear of readability issues. Ideally, I would love to use f-strings for better readability, but that change should happen everywhere at once, which will also up the minimum Python version requirement to 3.6+.

Type annotations are great, but they would mean we are not supporting anything before Python 3.5, right? Also, should we defer type annotations to apply globally at once in a separate PR rather than introducing one function looking different from the rest?

Reply:
1. A variable created once and once used is usually called an *explanatory variable*. Its point is to name a thing for the person, not for the computer, and show a human what is going on.

In this case, keyword arguments to `.format()` are indeed a bit too much - but I don't see they hurt, either :) I can remove them if you guys want more brevity.

> Type annotations are great, but they would mean we are not supporting anything before Python 3.5, right?

That's correct. However, I am not sure whether this is an issue since I am not sure in what environments this project is being used. It is not a library, so it can't break someone's project by not supporting an older Python version.

> Also, should we defer type annotations to apply globally at once in a separate PR rather than introducing one function looking different from the rest?

I do not believe so; there is a notion of "gradual typing" in Python world meaning a slow step by step addition of type hints into a Python project.

One of the reasons for this is that a PR to add all types everywhere AND make mypy happy would be too large. I prefer smaller PRs, easier to review and to think about.

Reply:
> I would be open to overhauling the naming if it puts us more inline with Python standards, which might in turn facilitate further contributions by those already familiar with the style.

If you create an issue, I will be happy to do something with it over this week. :)